File: src/seogeo/verification.py
from __future__ import annotations

"""Baseline, regression, and post-deploy verification helpers."""

import json
from dataclasses import dataclass
from pathlib import Path

from seogeo.models import Finding


@dataclass(slots=True)
class DiffResult:
    """Comparison result between baseline and current findings."""

    new_findings: list[Finding]
    resolved_findings: list[Finding]
    unchanged_findings: list[Finding]


def load_findings_from_audit(path: Path) -> list[Finding]:
    """Load persisted findings from an audit JSON file."""
    payload = json.loads(path.read_text())
    if not isinstance(payload, list):
        return []
    return [Finding.from_dict(item) for item in payload if isinstance(item, dict)]


def write_baseline_file(findings: list[Finding], path: Path) -> Path:
    """Persist a finding set as a reusable regression baseline."""
    path.write_text(json.dumps([finding.to_dict() for finding in findings], indent=2))
    return path


def diff_finding_sets(baseline: list[Finding], current: list[Finding]) -> DiffResult:
    """Return new, resolved, and unchanged findings using stable finding fingerprints."""
    baseline_by_key = {finding.fingerprint(): finding for finding in baseline}
    current_by_key = {finding.fingerprint(): finding for finding in current}
    new_keys = sorted(set(current_by_key) - set(baseline_by_key))
    resolved_keys = sorted(set(baseline_by_key) - set(current_by_key))
    unchanged_keys = sorted(set(current_by_key) & set(baseline_by_key))
    return DiffResult(
        new_findings=[current_by_key[key] for key in new_keys],
        resolved_findings=[baseline_by_key[key] for key in resolved_keys],
        unchanged_findings=[current_by_key[key] for key in unchanged_keys],
    )


def render_diff_text(diff: DiffResult) -> str:
    """Render a human-readable regression summary."""
    lines = [
        "Diff Report",
        "",
        f"New findings: {len(diff.new_findings)}",
        f"Resolved findings: {len(diff.resolved_findings)}",
        f"Unchanged findings: {len(diff.unchanged_findings)}",
    ]
    if diff.new_findings:
        lines.extend(["", "New"])
        lines.extend(f"- {finding.render()}" for finding in diff.new_findings)
    if diff.resolved_findings:
        lines.extend(["", "Resolved"])
        lines.extend(f"- {finding.render()}" for finding in diff.resolved_findings)
    return "\n".join(lines)

File: crates/seogeo-core/src/verification.rs
use anyhow::Result;
use serde::Serialize;
use std::collections::BTreeMap;
use std::fs;
use std::path::Path;

use seogeo_contracts::{AuditArtifact, AuditStatus, Finding, FindingFingerprint};

use crate::reporting::{build_audit_artifact, summarize_findings};

#[derive(Debug, Clone, PartialEq, Eq, Serialize)]
pub struct DiffResult {
    pub new_findings: Vec<Finding>,
    pub resolved_findings: Vec<Finding>,
    pub unchanged_findings: Vec<Finding>,
}

pub fn load_audit_artifact(path: &Path) -> Result<AuditArtifact> {
    let text = fs::read_to_string(path)?;
    if let Ok(artifact) = serde_json::from_str::<AuditArtifact>(&text) {
        return Ok(artifact);
    }
    let findings = serde_json::from_str::<Vec<Finding>>(&text)?;
    let mut artifact = build_audit_artifact("legacy", &findings, AuditStatus::Complete, None, None);
    artifact.generated_at = 0;
    artifact.summary = summarize_findings(&artifact.findings);
    Ok(artifact)
}

pub fn load_findings_from_audit(path: &Path) -> Result<Vec<Finding>> {
    Ok(load_audit_artifact(path)?.findings)
}

pub fn write_baseline_file(findings: &[Finding], path: &Path) -> Result<()> {
    let artifact = build_audit_artifact("baseline", findings, AuditStatus::Complete, None, None);
    fs::write(path, serde_json::to_string_pretty(&artifact)?)?;
    Ok(())
}

pub fn diff_finding_sets(baseline: &[Finding], current: &[Finding]) -> DiffResult {
    let baseline_by_key: BTreeMap<FindingFingerprint, Finding> = baseline
        .iter()
        .cloned()
        .map(|finding| (finding.fingerprint(), finding))
        .collect();
    let current_by_key: BTreeMap<FindingFingerprint, Finding> = current
        .iter()
        .cloned()
        .map(|finding| (finding.fingerprint(), finding))
        .collect();

    let new_findings = current_by_key
        .iter()
        .filter(|(key, _)| !baseline_by_key.contains_key(key))
        .map(|(_, finding)| finding.clone())
        .collect();
    let resolved_findings = baseline_by_key
        .iter()
        .filter(|(key, _)| !current_by_key.contains_key(key))
        .map(|(_, finding)| finding.clone())
        .collect();
    let unchanged_findings = current_by_key
        .iter()
        .filter(|(key, _)| baseline_by_key.contains_key(key))
        .map(|(_, finding)| finding.clone())
        .collect();

    DiffResult {
        new_findings,
        resolved_findings,
        unchanged_findings,
    }
}

pub fn render_diff_text(diff: &DiffResult) -> String {
    let mut lines = vec![
        "Diff Report".to_string(),
        String::new(),
        format!("New findings: {}", diff.new_findings.len()),
        format!("Resolved findings: {}", diff.resolved_findings.len()),
        format!("Unchanged findings: {}", diff.unchanged_findings.len()),
    ];
    if !diff.new_findings.is_empty() {
        lines.push(String::new());
        lines.push("New".to_string());
        for finding in &diff.new_findings {
            lines.push(format!("- {}", finding.render()));
        }
    }
    if !diff.resolved_findings.is_empty() {
        lines.push(String::new());
        lines.push("Resolved".to_string());
        for finding in &diff.resolved_findings {
            lines.push(format!("- {}", finding.render()));
        }
    }
    lines.join("\n")
}

#[cfg(test)]
mod tests {
    use super::{diff_finding_sets, load_audit_artifact, render_diff_text, write_baseline_file};
    use seogeo_contracts::{AuditStatus, Finding, FindingScope};

    fn finding(rule_id: &str, path: &str) -> Finding {
        Finding {
            rule_id: rule_id.into(),
            message: "msg".into(),
            path: path.into(),
            line: 1,
            column: 1,
            severity: "error".into(),
            suggestion: None,
            scope: FindingScope::Page,
        }
    }

    #[test]
    fn diffs_findings_by_fingerprint() {
        let baseline = vec![finding("SEO001", "a.html")];
        let current = vec![finding("SEO001", "a.html"), finding("SEO002", "b.html")];
        let diff = diff_finding_sets(&baseline, &current);
        assert_eq!(diff.new_findings.len(), 1);
        assert_eq!(diff.unchanged_findings.len(), 1);
        assert!(render_diff_text(&diff).contains("New findings: 1"));
    }

    #[test]
    fn loads_legacy_findings_arrays() {
        let temp_dir = tempfile::tempdir().unwrap();
        let path = temp_dir.path().join("legacy.json");
        std::fs::write(
            &path,
            serde_json::to_string_pretty(&vec![finding("SEO001", "a.html")]).unwrap(),
        )
        .unwrap();
        let artifact = load_audit_artifact(&path).unwrap();
        assert_eq!(artifact.status, AuditStatus::Complete);
        assert_eq!(artifact.findings.len(), 1);
    }

    #[test]
    fn writes_structured_baseline_files() {
        let temp_dir = tempfile::tempdir().unwrap();
        let path = temp_dir.path().join("baseline.json");
        write_baseline_file(&[finding("SEO001", "a.html")], &path).unwrap();
        let artifact = load_audit_artifact(&path).unwrap();
        assert_eq!(artifact.command, "baseline");
        assert_eq!(artifact.findings.len(), 1);
    }
}
